lib/services/local/saml_idp_service_provider.go

```diff
@@ -295,3 +306,82 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
 	sp.SetEntityDescriptor(string(entityDescriptor))
 	return nil
 }
+
+// embedAttributeMapping embeds attribute mapping input into entity descriptor.
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	existingTeleportSPSSODescriptorIndex, _ := GetTeleportSPSSODescriptor(ed.SPSSODescriptors)
+	switch atrrMapLen := len(sp.GetAttributeMapping()); {
+	case atrrMapLen == 0 && existingTeleportSPSSODescriptorIndex == 0:
+		s.log.Debugf("no custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
+		return nil
+	case atrrMapLen == 0 && existingTeleportSPSSODescriptorIndex > 0:
+		// delete Teleport SPSSODescriptor
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors[:existingTeleportSPSSODescriptorIndex], ed.SPSSODescriptors[existingTeleportSPSSODescriptorIndex+1:]...)
+	case atrrMapLen > 0 && existingTeleportSPSSODescriptorIndex == 0:
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors, genTeleportSPSSODescriptor(sp.GetAttributeMapping()))
+	case atrrMapLen > 0 && existingTeleportSPSSODescriptorIndex > 0:
+		// if there is existing SPSSODescriptor with "TELEPORT_SAML_IDP" service name, replace it at
+		// existingTeleportSPSSODescriptorIndex to avoid duplication or possible fragmented SPSSODescriptor.
+		ed.SPSSODescriptors[existingTeleportSPSSODescriptorIndex] = genTeleportSPSSODescriptor(sp.GetAttributeMapping())
+	}
```

Comment:
I think switch statement hurts readability here. How about a simple "if" with comments explaining what each case is?

```go
attrMapLen := len(sp.GetAttributeMapping())

if attrMapLen == 0 { // Service provider doesn't have custom attribute mapping configured.
  if existingTeleportSPSSODescriptorIndex == 0 {     // No existing Teleport descriptor, nothing to do.
    return nil
  } else {     // Descriptor has Teleport attributes, remove it.
    ...
  }
} else { // Service provider has custom attributes mapping
  if existingTeleportSPSSODescriptorIndex == 0 {     // Teleport descriptor doesn't exist, add it.
    ...
  } else {     // Teleport descriptor already exists, update it.
    ...
  }
}
```

Reply:
Yeah right. But not really fan of deep nested if else either. I think the culprit is `existingTeleportSPSSODescriptorIndex` which actually can be `teleportSPSSODescriptorIndex` since this is the only value in `embedAttributeMapping()`.
I came up with this one, looks a bit better than previous:
```
switch attrMapLen := len(sp.GetAttributeMapping()); {
	case attrMapLen == 0:
		if teleportSPSSODescriptorIndex == 0 {
			s.log.Debugf("No custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
		} else {
			// delete Teleport SPSSODescriptor
			ed.SPSSODescriptors = append(ed.SPSSODescriptors[:teleportSPSSODescriptorIndex], ed.SPSSODescriptors[teleportSPSSODescriptorIndex+1:]...)
		}
	case attrMapLen > 0:
		if teleportSPSSODescriptorIndex == 0 {
			ed.SPSSODescriptors = append(ed.SPSSODescriptors, genTeleportSPSSODescriptor(sp.GetAttributeMapping()))
		} else {
			// if there is existing SPSSODescriptor with "teleport_saml_idp_service" service name, replace it at
			// existingTeleportSPSSODescriptorIndex to avoid duplication or possible fragmented SPSSODescriptor.
			ed.SPSSODescriptors[teleportSPSSODescriptorIndex] = genTeleportSPSSODescriptor(sp.GetAttributeMapping())
		}
	}
```